### PR TITLE
🎉 Update destination-s3 to handle the new data types protocol

### LINKED ISSUE #17645: Update destination-s3 to handle the new data types protocol
See https://github.com/airbytehq/airbyte/issues/17637.

Definition of done:
 - destination-s3 accepts the new protocol messages
 - destination data does not change when protocol message version changes (should write a test for this)

### PR DESCRIPTION
**Pull request by suhomud**
## What
New version protocol support V1. See https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/well_known_types.yaml

## How
In order to support new protocol the base-java-s3 module should handle new json schema types: 
```
   STRING_V1("WellKnownTypes.json#/definitions/String", Schema.Type.STRING),
   INTEGER_V1("WellKnownTypes.json#/definitions/Integer", Schema.Type.INT),
   NUMBER_V1("WellKnownTypes.json#/definitions/Number", Schema.Type.DOUBLE),
   BOOLEAN_V1("WellKnownTypes.json#/definitions/Boolean", Schema.Type.BOOLEAN),
   BINARY_DATA_V1("WellKnownTypes.json#/definitions/BinaryData", Schema.Type.BYTES),
   DATE_V1("WellKnownTypes.json#/definitions/Date", Schema.Type.INT),
   TIMESTAMP_WITH_TIMEZONE_V1("WellKnownTypes.json#/definitions/TimestampWithTimezone", Schema.Type.LONG),
   TIMESTAMP_WITHOUT_TIMEZONE_V1("WellKnownTypes.json#/definitions/TimestampWithoutTimezone", Schema.Type.LONG),
   TIME_WITH_TIMEZONE_V1("WellKnownTypes.json#/definitions/TimeWithTimezone", Schema.Type.LONG),
   TIME_WITHOUT_TIMEZONE_V1("WellKnownTypes.json#/definitions/TimeWithoutTimezone", Schema.Type.LONG),
```
V0 protocol is still supported but marked as `@Deprecated`:
```
  @Deprecated
   STRING_V0("string", null, Schema.Type.STRING),
   @Deprecated
   NUMBER_INT_V0("number", "integer", Schema.Type.INT),
   @Deprecated
   NUMBER_BIGINT_V0("string", "big_integer", Schema.Type.STRING),
   @Deprecated
   NUMBER_FLOAT_V0("number", "float", Schema.Type.FLOAT),
   @Deprecated
   NUMBER_V0("number", null, Schema.Type.DOUBLE),
   @Deprecated
   INTEGER_V0("integer", null, Schema.Type.INT),
   @Deprecated
   BOOLEAN_V0("boolean", null, Schema.Type.BOOLEAN),
   @Deprecated
   NULL("null", null, Schema.Type.NULL);
``` 

This PR [Support era and string numbers](https://github.com/airbytehq/json-avro-converter/pull/17)  for  [json-aero-converter ](https://github.com/airbytehq/json-avro-converter) adding the string numbers and new Date type support. Released version is 1.0.2

## Recommended reading order
1. `JsonSchemaType.java` new datatypes 
2. `JsonToAvroSchemaConverter.java` build avro schema with new protocol. 
Test related changes: 
3. ` AdvancedTestDataComparator.java` Updated regex to match new date format. Fixed  `isDateTimeWithTzValue` since it was not working and was not able to match date time with timezone. New `compareTime` method. Added workoround `TEST_DATASET_IGNORE_LIST` those values are used as string value in the dataset i.e. in `exchange_rate_messages.txt` and cannot pass some tests (probably need to be re-visit)
4. `S3AvroParquetTestDataComparator.java` new S3 specific test data comparator (the old one renamed to `S3BaseAvroParquetTestDataComparator`)
5. S3 destination integration test now return `ProtocolVersion.V1` v1 test dataset
## 🚨 User Impact 🚨
The only noticable thing is when Avro schema has `[int, string] ` types it will try to produce number from the string so "123" become 123 in result Avro object. 

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `Airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `Airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java
Comment (grishick) [inline airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java:37]: does this comment imply that if we merge this change to master, we will see new failures when bigquery-denormalized is trying to process datetime values? Or will this still be OK until bigquery-denormalized declares that it supports v1 protocol?

**Review by grishick — COMMENTED** [review]
Looks good overall. I added one question, but will defer to @edgao for approval

**Review by edgao — COMMENTED** [review]
how hard would it be to add a parameter to only use v0/v1 mode? Right now the logic for both is mixed together, so when we (eventually) migrate everything to v1, it'll be hard to remove the v0 logic.

Then destination connectors would explicitly need to be upgraded to use v1. Which also limits the blast radius in case there's a hidden bug here, since they're not going to mysteriously have new behavior after getting published.

File: airbyte-integrations/bases/base-java-s3/build.gradle
Comment (suhomud) [reply airbyte-integrations/bases/base-java-s3/build.gradle:15]: Agree. I'll release 1.1.0 from master branch as soon as we merge https://github.com/airbytehq/json-avro-converter/pull/17

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java
Comment (suhomud) [reply airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java:37]: Yeah, there were two issues here. 
1. Previous test comparison had a wrong regex to find date time with Tz values:
```
  protected boolean isDateTimeWithTzValue(final String value) {
    return value.matches(".+[+-]\\d{2}:\\d{2}");
  }
```
I had to fix it to handle it for avro values (it used microseconds). 
2. After this fix I got couple of dates which schema defined as a string but not date (dataset V0, V1). Tried to update schema  to: 
```
              "type": "string",
              "format": "date-time",
              "airbyte_type": "timestamp_with_timezone"
```
but  bigquery-denormalized failed to store `2020-08-31T00:00:00Z`.  I was not sure how this change affect tests for others connectors also, so created this workaround 

Created the issue https://github.com/airbytehq/airbyte/issues/20362. Don't have a logs now but will try to extend with details

File: airbyte-integrations/bases/standard-destination-test/src/main/resources/v1/data_type_array_test_catalog.json
Comment (suhomud) [reply airbyte-integrations/bases/standard-destination-test/src/main/resources/v1/data_type_array_test_catalog.json:39]: Totally agree, I'll remove it.

File: airbyte-integrations/bases/standard-destination-test/src/main/resources/v1/exchange_rate_catalog.json
Comment (suhomud) [reply airbyte-integrations/bases/standard-destination-test/src/main/resources/v1/exchange_rate_catalog.json:14]: Long story short :). This dataset is used in a different tests (see `EXCHANGE_RATE_CONFIG` constant  in a `DestinationAcceptanceTest`). One of the issue that I faced was testLineBreakCharacters test which broke avro converter   (which is why I added https://github.com/airbytehq/json-avro-converter/pull/17/files#diff-c994d6172cabf7dca250e0806b69c501d475caac548cfd9a574e2458b7da3b5aR24 by the way 😃) and I was not sure what will happen with others connector's tests if I'll change it. 

I see two variants:
1. keep `"$ref": "WellKnownTypes.json#/definitions/Date"`. Here I'll need to be sure that all it passed for destination s3 V1. Then all who add V1 support need to be sure that connector supports different variants of date time with timezone
2. Temporary use it as string and create a ticket for this refactoring. (I didn't notice it during V1 tests adaptation)

WDYT?

File: airbyte-integrations/connectors/destination-s3/src/test-integration/java/io/airbyte/integrations/destination/s3/S3AvroParquetTestDataComparator.java
Comment (suhomud) [reply airbyte-integrations/connectors/destination-s3/src/test-integration/java/io/airbyte/integrations/destination/s3/S3AvroParquetTestDataComparator.java:26]: It has both microseconds and millisecond
https://avro.apache.org/docs/1.11.0/spec.html
Maybe we can migrate to milliseconds only when V1 is fully integrated? 

Yes, `Long.parseLong(epochValue) / 1000` is better 👍

**Comment by suhomud** [comment]
> how hard would it be to add a parameter to only use v0/v1 mode? Right now the logic for both is mixed together, so when we (eventually) migrate everything to v1, it'll be hard to remove the v0 logic.
> 
> Then destination connectors would explicitly need to be upgraded to use v1. Which also limits the blast radius in case there's a hidden bug here, since they're not going to mysteriously have new behavior after getting published.

Not sure that I fully understand (probably let's sync about it). Generally speaking the only adaptation that I see for removal of V0 logic is to remove old schema json to avro types https://github.com/airbytehq/airbyte/pull/20088/files#diff-ccb0476def819bb3177cf9458fb6939bc25c42ea516305171e1e11be2a3d1a39R32 (It's marked as deprecated) Plus some places here https://github.com/airbytehq/airbyte/pull/20088/files#diff-d25b772db4efd1799ea8a085d4acc27f7f5393fda2dca2498c2d27391e6148deR235 
The `json-aero-converter` is not required adaptation

File: airbyte-integrations/bases/standard-destination-test/src/main/resources/v1/exchange_rate_catalog.json
Comment (suhomud) [reply airbyte-integrations/bases/standard-destination-test/src/main/resources/v1/exchange_rate_catalog.json:14]: Agree with you but probably will need to revisit it. Since now it will a bit of pain to adapt it. 
The data types are tested with `data_type_object_test_catalog.json, data_type_array_test_catalog.json, data_type_array_object_test_catalog.json` datasets

**Review by edgao — APPROVED** [review]
🚛

**Comment by edgao** [comment]
@suhomud - I did a test with some more complex data types and I think there's some (maybe?) slightly wrong behavior with union types. Specifically, if you have a schema with `oneOf: [{$ref: String}, {$ref: Integer}]`, it will always become an avro string. But If you have `oneOf: [Integer, String]` then it will parse integers when possible, and fall back to string the rest of the time. Similar behavior for e.g. an array's `items` schema.

Though... we get the same thing with timestamps, which is maybe(?) preexisting behavior. So maybe we're just OK with this?

Test procedure:
1. Create a directory `~/Desktop/protocol_test`
2. Write three files into that directory:
    1. `catalog.json`
       ```json
       {
         "streams": [
           {
             "stream": {
               "name": "protocol_test",
               "namespace": "public",
               "json_schema": {
                 "type": "object",
                 "properties": {
                   "id": {"$ref": "WellKnownTypes.json#/definitions/Integer"},
                   "str_int_data": {
                     "oneOf": [
                       {"$ref": "WellKnownTypes.json#/definitions/String"},
                       {"$ref": "WellKnownTypes.json#/definitions/Integer"}
                     ]
                   },
                   "str_timestamp_data": {
                     "oneOf": [
                       {"$ref": "WellKnownTypes.json#/definitions/String"},
                       {"$ref": "WellKnownTypes.json#/definitions/TimestampWithTimezone"}
                     ]
                   }
                 }
               },
               "default_cursor_field": [],
               "supported_sync_modes": [
                 "full_refresh",
                 "incremental"
               ],
               "source_defined_primary_key": [["id"]]
             },
             "sync_mode": "full_refresh",
             "primary_key": [["id"]],
             "cursor_field": [],
             "destination_sync_mode": "overwrite"
           }
         ]
       }
       ```
    2. `config.json` (replace access_key_id and secret_access_key with real values from https://console.cloud.google.com/security/secret-manager/secret/SECRET_DESTINATION-S3_CREDS/versions?project=dataline-integration-testing)
       ```json
       {
         "format": {
           "format_type": "Avro",
           "compression_codec": {
             "codec": "no compression"
           }
         },
         "s3_endpoint": "",
         "access_key_id": "<redacted>",
         "s3_bucket_name": "airbyte-integration-test-destination-s3",
         "s3_bucket_path": "edgao_test",
         "s3_bucket_region": "us-east-2",
         "secret_access_key": "<redacted>"
       }
       ```
    3. `inp.jsonl`
       ```json
       {"type": "RECORD", "record": {"stream": "protocol_test", "namespace": "public", "emitted_at": 1602637589000, "data": {"id": "1","str_int_data": "42", "str_timestamp_data": "2022-01-23T12:34:56.789Z"}}}
       ```
3. `cat ~/Desktop/protocol_test/inp.jsonl | docker run -i -v ~/Desktop/protocol_test:/in airbyte/destination-s3:dev write --catalog /in/catalog.json --config /in/config.json`
4. Download https://dlcdn.apache.org/avro/avro-1.11.1/java/avro-tools-1.11.1.jar to somewhere
5. Download the avro file (e.g. `aws s3 cp s3://airbyte-integration-test-destination-s3/edgao_test/public/protocol_test/2022_12_23_1671832609043_0.avro ~/Desktop/replicated.avro`)
6. `java -jar path/to/avro-tools.jar tojson ~/Desktop/replicated.avro` (you can also use `java -jar path/to/avro-tools.jar getschema ~/Desktop/replicated.avro` to see the avro schema)

Observe that str_int_data and str_timestamp_data both got converted to avro strings. But if you edit catalog.json to swap the order of the `oneOf`, then they get converted to int/long.

**Comment by suhomud** [comment]
@edgao great analysis! I think this behaviour was here before this PR and it is related to json-avro-converter lib. So if the first type in union is a string then it checks if it is an instance of string. In our case it is true, therefor nothing prevent converter to put it as a string because it is a valid string

We can think about making `readUnion` more intelligent one the  json-avro-converter side. What I can see is to make some parsing order for example if union has `TimestampWithTimezone` and `String` types then first it will "try" to parse `TimestampWithTimezone` and if failed `String`. More complex example is numbers since user can has a data with numbers stored as string so here we can still have a mismatch

**Comment by edgao** [comment]
nice, makes sense. Let's just keep this behavior for now.

**Pull request merged**
